### PR TITLE
GetOrthogonal

### PR DESCRIPTION
**Pull request by seleanridley**

### CODE REVIEW

File: cpp/include/cpu_operations.h [hunk 1]
```diff
@@ -400,6 +401,29 @@ class CpuOperations {
      exit(1);
     }
 }
+  static Vector<T> GetOrthogonal(const Vector<T> &a) {
+    if ( a.mean() == 0 )
```
Comment (yiskylee) [inline cpp/include/cpu_operations.h]: No spaces inside of the parentheses

Comment (yiskylee) [inline cpp/include/cpu_operations.h]: a.mean() == 0 does not mean that all of the elements are zeros

File: cpp/include/cpu_operations.h [hunk 2]
```diff
@@ -400,6 +401,29 @@ class CpuOperations {
...
+     std::cerr << "Vector can not be all zeros!";
+    Vector<T> b(a.size());
+    b.setOnes();
+    Vector<T> projection;
+    projection = ((a.dot(b))/(pow(a.norm(), 2))) * a;
```
Comment (yiskylee) [inline cpp/include/cpu_operations.h:507]: a.norm() returns sqrt(|a|^2), a.squarednorm() returns |a|^2

**Comment by yiskylee** [comment]
@seleanridley , I realized that we need to refactor the getOrthogonal just a little bit, let me know what time you could meet to discuss. Thanks!

File: cpp/include/cpu_operations.h (hunk 2 above)
Comment (seleanridley) [inline cpp/include/cpu_operations.h]: Are you referring to norm() or squaredNorm() ?

Comment (yiskylee) [inline cpp/include/cpu_operations.h]: a.norm() returns |a|^2 and squaredNorm() returns sqrt(|a|^2)

**Comment by yiskylee** [comment]
@seleanridley , could you make another pull from upstream?

**Comment by yiskylee** [comment]
Hi @seleanridley , I did not notice this before but I think we might need to make a change to the output parameter to GetOrthogonal: Vector<T> &b. According to the google code style: https://google.github.io/styleguide/cppguide.html#Reference_Arguments
They prefer to use pointer instead of non-const reference for parameters that are output. So we want to change Vector<T> &b to Vector<T> *b, and change things like b -= projection to *b -= projection (where \* is used to de-reference the pointer b). You will also need to make changes to your get orthogonal test to pass in something like &b where your b is a Vector<T> and "&" is called "address-of" operator which returns a pointer pointed to b. 
You can make a pull from upstream develop to update your branch first, and then you can see examples of how to do this change of parameters in my updated GenDegreeMatrix() in cpu_operations.h

**Pull request closed without merging**